### PR TITLE
Optimize GreenNode.CreateList

### PR DESCRIPTION
**Pull request by HurricanKai**
This PR improves the performance of GreenNode.CreateList.
This is part of a larger goal to improve NormalizeWhitespace performance, the benchmark below tests NormalizeWhitespace performance of real-world generated C# source (~116k loc when normalized).
Benchmark:
|   Method |    Mean |    Error |   StdDev |      Gen 0 |      Gen 1 | Gen 2 | Allocated |
|--------- |--------:|---------:|---------:|-----------:|-----------:|------:|----------:|
| Baseline | 5.806 s | 0.0556 s | 0.0493 s | 74000.0000 | 18000.0000 |     - | 591.82 MB |
| New      | 5.595 s | 0.1072 s | 0.1731 s | 49000.0000 | 13000.0000 |     - | 393.68 MB |

Tested on Net5.0 RC2 with the Benchmark runtime being .NET Core 3.1.8

I also investigated a couple of other approaches (ImmutableArray + SelectAsArray, GetEnumerator directly) but this one came out on top.
Thanks for the help @333fred and @CyrusNajmabadi 😄 

Benchmark Files (drop into IdeCoreBenchmarks for testing):
[normalization-sample.txt](https://github.com/dotnet/roslyn/files/5373857/normalization-sample.txt)
[NormalizeWhitespaceBenchmark.cs](https://github.com/dotnet/roslyn/files/5373859/NormalizeWhitespaceBenchmark.txt)

### CODE REVIEW

File: src/Compilers/Core/Portable/Syntax/GreenNode.cs [hunk 1]
```diff
@@ -924,17 +922,22 @@ private static bool EquivalentToInternal(GreenNode node1, GreenNode node2)
...
+                {
+                    var array = new ArrayElement<GreenNode>[list.Count];
+                    for (int i = 0; i < array.Length; i++)
+                        array[i].Value = select(list[i]);
+                    return SyntaxList.List(array);
+                }
             }
         }
-
+
```
Comment (333fred) [inline src/Compilers/Core/Portable/Syntax/GreenNode.cs:940]: Trailing whitespace, the formatter leg will complain.

File: src/Compilers/Core/Portable/Syntax/SyntaxToken.cs [hunk 2]
```diff
@@ -473,14 +473,18 @@ public SyntaxToken WithLeadingTrivia(params SyntaxTrivia[]? trivia)
...
 
         /// <summary>
-        /// Creates a new token from this token with the leading trivia specified..
+        /// Creates a new token from this token with the leading trivia specified.
         /// </summary>
         public SyntaxToken WithLeadingTrivia(IEnumerable<SyntaxTrivia>? trivia)
         {
-            var greenList = trivia?.Select(t => t.RequiredUnderlyingNode);
-
+            IReadOnlyList<SyntaxTrivia>? list = trivia switch
```
Comment (333fred) [inline src/Compilers/Core/Portable/Syntax/SyntaxToken.cs:480]: Consider extracting to a helper, shared with the code below.

File: src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs [hunk 3]
```diff
@@ -315,7 +315,7 @@ private static SyntaxList<TNode> CreateList(List<TNode> items)
                 return default(SyntaxList<TNode>);
             }
 
-            var newGreen = GreenNode.CreateList(items.Select(n => n.Green));
+            var newGreen = GreenNode.CreateList(items, n => n.Green);
```
Comment (333fred) [inline src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs:318]:
```suggestion
            var newGreen = GreenNode.CreateList(items, static n => n.Green);
```

Consider doing this for all the lambdas in this PR to make it very obvious they can't capture and are cacheable.

File: src/Compilers/Core/Portable/Syntax/SyntaxNodeOrTokenList.cs [hunk 4]
```diff
@@ -316,10 +316,10 @@ private static SyntaxNodeOrTokenList CreateList(List<SyntaxNodeOrToken> items)
                 return default(SyntaxNodeOrTokenList);
             }
 
-            var newGreen = GreenNode.CreateList(items.Select(n => n.UnderlyingNode!))!;
+            var newGreen = GreenNode.CreateList(items, n => n.UnderlyingNode!)!;
             if (newGreen.IsToken)
             {
-                newGreen = GreenNode.CreateList(new[] { newGreen }, alwaysCreateListNode: true)!;
+                newGreen = Syntax.InternalSyntax.SyntaxList.List(new[] { newGreen });
```
Comment (333fred) [inline src/Compilers/Core/Portable/Syntax/SyntaxNodeOrTokenList.cs:322]: Why this change? Unless this is a particular hotspot, I'd prefer just passing an identity selector.

**Review by 333fred — CHANGES_REQUESTED** [review]
Overall looks good, a couple of issues.

File: src/Compilers/Core/Portable/Syntax/GreenNode.cs [hunk 5]
```diff
@@ -904,16 +904,14 @@ private static bool EquivalentToInternal(GreenNode node1, GreenNode node2)
         public abstract SyntaxToken CreateSeparator<TNode>(SyntaxNode element) where TNode : SyntaxNode;
         public abstract bool IsTriviaWithEndOfLine(); // trivia node has end of line
 
-        public static GreenNode? CreateList(IEnumerable<GreenNode>? nodes, bool alwaysCreateListNode = false)
+        public static GreenNode? CreateList<TFrom>(IReadOnlyList<TFrom>? list, Func<TFrom, GreenNode> select, bool alwaysCreateListNode = false)
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/GreenNode.cs:907]:
```suggestion
        public static GreenNode? CreateList<TFrom>(IReadOnlyList<TFrom>? list, Func<TFrom, GreenNode> select, bool alwaysCreateListNode = false)
```

curious if we have to take `list` in by interface, and not some concreate type.

File: src/Compilers/Core/Portable/Syntax/SyntaxTriviaList.cs [hunk 6]
```diff
@@ -397,7 +397,7 @@ public SyntaxTriviaList ReplaceRange(SyntaxTrivia triviaInList, IEnumerable<Synt
                 var list = this.ToList();
                 list.RemoveAt(index);
                 list.InsertRange(index, newTrivia);
-                return new SyntaxTriviaList(default(SyntaxToken), GreenNode.CreateList(list.Select(n => n.RequiredUnderlyingNode)), 0, 0);
+                return new SyntaxTriviaList(default(SyntaxToken), GreenNode.CreateList(list, n => n.RequiredUnderlyingNode), 0, 0);
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/SyntaxTriviaList.cs:400]: looks like mosts callers pass in a List.  is there a need to specify IReadOnlyList rather than optmizing for that concrete type?

File: src/Compilers/Core/Portable/Syntax/GreenNode.cs (hunk 5 above)
Comment (HurricanKai) [reply src/Compilers/Core/Portable/Syntax/GreenNode.cs:907]: Not necessarily, it could be List<TFrom> but that would make it much less likely for the efficient branch in the trivia methods to be hit.
And I'm not sure what the advantage would be

File: src/Compilers/Core/Portable/Syntax/SyntaxNodeOrTokenList.cs (hunk 4 above)
Comment (HurricanKai) [reply src/Compilers/Core/Portable/Syntax/SyntaxNodeOrTokenList.cs:322]: I prefered this because it allows me to remove the optional bool parameter. Should I revert this?

File: src/Compilers/Core/Portable/Syntax/SyntaxToken.cs [hunk 7]
```diff
@@ -472,19 +472,26 @@ public SyntaxToken WithLeadingTrivia(params SyntaxTrivia[]? trivia)
...
 
-        /// <summary>
-        /// Creates a new token from this token with the leading trivia specified.
-        /// </summary>
-        public SyntaxToken WithLeadingTrivia(IEnumerable<SyntaxTrivia>? trivia)
+        private static IReadOnlyList<SyntaxTrivia>? GetList(IEnumerable<SyntaxTrivia>? trivia)
         {
-            IReadOnlyList<SyntaxTrivia>? list = trivia switch
+            if (trivia is null)
+                return null;
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/SyntaxToken.cs:478]: this seems unnecessary given the switch expression handles null.

File: src/Compilers/Core/Portable/Syntax/SyntaxToken.cs [hunk 8]
```diff
@@ -472,19 +472,26 @@ public SyntaxToken WithLeadingTrivia(params SyntaxTrivia[]? trivia)
             return this.WithLeadingTrivia((IEnumerable<SyntaxTrivia>?)trivia);
         }
 
-        /// <summary>
-        /// Creates a new token from this token with the leading trivia specified.
-        /// </summary>
-        public SyntaxToken WithLeadingTrivia(IEnumerable<SyntaxTrivia>? trivia)
+        private static IReadOnlyList<SyntaxTrivia>? GetList(IEnumerable<SyntaxTrivia>? trivia)
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/SyntaxToken.cs:475]: AsReadOnlyList

File: src/Compilers/Core/Portable/Syntax/SyntaxNodeOrTokenList.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Syntax/SyntaxNodeOrTokenList.cs:322]: i'm good with this.

**Comment by HurricanKai** [comment]
Here are some performance figures. The currently committed version is `List` previously it was `IROList` and the current master is `Baseline` the others are for other alternatives I've considered.
The numbers aren't 100% accurate but should give a good idea what's fastest. Sorted by speed.
|     Method |    Mean |    Error |   StdDev |      Gen 0 |      Gen 1 | Gen 2 | Allocated |
|----------- |--------:|---------:|---------:|-----------:|-----------:|------:|----------:|
| List       | 5.683 s | 0.0338 s | 0.0316 s | 72000.0000 | 18000.0000 |     - | 579.89 MB |
| List/Array | 5.861 s | 0.0505 s | 0.0472 s | 67000.0000 | 17000.0000 |     - | 536.25 MB |
| IROList    | 5.889 s | 0.0371 s | 0.0310 s | 49000.0000 | 13000.0000 |     - | 393.68 MB |
| Baseline   | 5.912 s | 0.0356 s | 0.0316 s | 74000.0000 | 18000.0000 |     - | 591.82 MB |

**Comment by CyrusNajmabadi** [comment]
Based on those numbers, looks like we should stick with IROL and not use List.  Or, we have an overload for both or something (but htat seems like a lot of complexity).

**Comment by 333fred** [comment]
Is the `List<T>` version generating more garbage from the non-List paths? Looking at those numbers, while using IROL list a clear win in terms of memory, it doesn't actually make a different CPU-time-wise, as the IROL and Baseline are within the Error margin. I think I'd prefer to have both a List and IROL overload. The method complexity is not very high, and 200 MB of memory savings on a large `NormalizeWhitespace`, while nice, is basically peanuts for our scenarios.

**Comment by HurricanKai** [comment]
The only types every passed in are `List<T>` and `T[]` I've tested two overloads with those two, see `List/Array`, I'll test List/IROList too, but I'd be surprised if it yielded better/much different results. Maybe because most type passed to WithTrivia are IROList 🤔

**Comment by HurricanKai** [comment]
I've just tested this but the results aren't beyond some error, so I doubt it makes a difference.

**Comment by 333fred** [comment]
> I've just tested this

What did you test?

**Comment by HurricanKai** [comment]
Having both an overload for List & IROList
This is only useful in the WithTrivia case, so there the code I chose was
```cs
null => null,
List<TFrom> list => GreenNode.CreateList(list, select),
IReadOnlyList list => GreenNode.CreateList(list, select),
_ => GreenNode.CreateList(enumerable.ToList(), select)
```

File: src/Compilers/Core/Portable/Syntax/GreenNode.cs [hunk 9]
```diff
@@ -904,34 +904,56 @@ private static bool EquivalentToInternal(GreenNode node1, GreenNode node2)
...
+                    return SyntaxList.List(array);
+                }
             }
-
-            var list = nodes.ToArray();
-
-            switch (list.Length)
+        }
+        
+        public static GreenNode? CreateList<TFrom>(IReadOnlyList<TFrom> list, Func<TFrom, GreenNode> select)
```
Comment (333fred) [inline src/Compilers/Core/Portable/Syntax/GreenNode.cs:938]: Please leave a comment explaining why we have 2 overloads.

**Review by 333fred — COMMENTED** [review]
Just one more comment

**Review by 333fred — APPROVED** [review]
LGTM (commit 9). @dotnet/roslyn-compiler for a second review of this community perf improvement.

**Comment by HurricanKai** [comment]
I've updated the original post to show current numbers.

**Review by cston — APPROVED** [review]

File: src/Compilers/Core/Portable/Syntax/SyntaxTokenList.cs [hunk 10]
```diff
@@ -377,7 +377,7 @@ public SyntaxTokenList RemoveAt(int index)
 
             var list = this.ToList();
             list.RemoveAt(index);
-            return new SyntaxTokenList(null, GreenNode.CreateList(list.Select(n => n.Node!)), 0, 0);
+            return new SyntaxTokenList(null, GreenNode.CreateList(list, static n => n.Node!), 0, 0);
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Syntax/SyntaxTokenList.cs:380]:
```suggestion
            return new SyntaxTokenList(null, GreenNode.CreateList(list, static n => n.RequiredNode), 0, 0);
```

File: src/Compilers/Core/Portable/Syntax/SyntaxTokenList.cs [hunk 11]
```diff
@@ -423,7 +423,7 @@ public SyntaxTokenList ReplaceRange(SyntaxToken tokenInList, IEnumerable<SyntaxT
                 var list = this.ToList();
                 list.RemoveAt(index);
                 list.InsertRange(index, newTokens);
-                return new SyntaxTokenList(null, GreenNode.CreateList(list.Select(n => n.Node!)), 0, 0);
+                return new SyntaxTokenList(null, GreenNode.CreateList(list, static n => n.Node!), 0, 0);
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Syntax/SyntaxTokenList.cs:426]:
```suggestion
                return new SyntaxTokenList(null, GreenNode.CreateList(list, static n => n.RequiredNode), 0, 0);
```

File: src/Compilers/Core/Portable/Syntax/SyntaxNodeOrTokenList.cs [hunk 12]
```diff
@@ -316,10 +316,10 @@ private static SyntaxNodeOrTokenList CreateList(List<SyntaxNodeOrToken> items)
                 return default(SyntaxNodeOrTokenList);
             }
 
-            var newGreen = GreenNode.CreateList(items.Select(n => n.UnderlyingNode!))!;
+            var newGreen = GreenNode.CreateList(items, static n => n.UnderlyingNode!)!;
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Syntax/SyntaxNodeOrTokenList.cs:319]:
```suggestion
            var newGreen = GreenNode.CreateList(items, static n => n.RequiredUnderlyingNode)!;
```

File: src/Compilers/Core/Portable/Syntax/SyntaxNodeOrTokenList.cs [hunk 13]
```diff
@@ -316,10 +316,10 @@ private static SyntaxNodeOrTokenList CreateList(List<SyntaxNodeOrToken> items)
...
             if (newGreen.IsToken)
             {
-                newGreen = GreenNode.CreateList(new[] { newGreen }, alwaysCreateListNode: true)!;
+                newGreen = Syntax.InternalSyntax.SyntaxList.List(new[] { newGreen });
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Syntax/SyntaxNodeOrTokenList.cs:322]: 💡 Use the overload that takes `ArrayElement<GreenNode>[]` to avoid an intermediate `GreenNode[1]` allocation.

File: src/Compilers/Core/Portable/Syntax/GreenNode.cs [hunk 14]
```diff
@@ -904,34 +904,61 @@ private static bool EquivalentToInternal(GreenNode node1, GreenNode node2)
...
+                    return SyntaxList.List(select(list[0]), select(list[1]));
                 case 3:
-                    return SyntaxList.List(list[0], list[1], list[2]);
-                default:
-                    return SyntaxList.List(list);
+                    return SyntaxList.List(select(list[0]), select(list[1]), select(list[2]));
+                default: {
+                    var array = new ArrayElement<GreenNode>[list.Count];
+                    for (int i = 0; i < array.Length; i++)
+                        array[i].Value = select(list[i]);
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Syntax/GreenNode.cs:959]: 📝 Even though `list` was already declared as having non-null `GreenNode` items, the new behavior here is slightly different for the case where a null value appeared. I'm not sure if that's a problem or not.

File: src/Compilers/Core/Portable/Syntax/GreenNode.cs [hunk 15]
```diff
@@ -904,34 +904,61 @@ private static bool EquivalentToInternal(GreenNode node1, GreenNode node2)
...
+                    return SyntaxList.List(array);
+                }
             }
+        }
 
-            var list = nodes.ToArray();
-
-            switch (list.Length)
+        public static GreenNode? CreateList<TFrom>(IReadOnlyList<TFrom> list, Func<TFrom, GreenNode> select)
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Syntax/GreenNode.cs:944]: 💭 This does make me wonder if we could somehow change this to only have one overload that takes `ReadOnlySpan<TFrom>`. Perhaps for the future...

File: src/Compilers/Core/Portable/Syntax/GreenNode.cs [hunk 16]
```diff
@@ -904,34 +904,61 @@ private static bool EquivalentToInternal(GreenNode node1, GreenNode node2)
         public abstract SyntaxToken CreateSeparator<TNode>(SyntaxNode element) where TNode : SyntaxNode;
         public abstract bool IsTriviaWithEndOfLine(); // trivia node has end of line
 
-        public static GreenNode? CreateList(IEnumerable<GreenNode>? nodes, bool alwaysCreateListNode = false)
+        /*
+         * There are 3 overloads of this, because most callers already know what they have is a List<T> and only transform it.
+         * In those cases List<TFrom> performs much better.
+         * In other cases, the type is unknown / is IEnumerable<T>, where we try to find the best match.
+         * There is another overload for IROList, since most collections already implement this, so checking for it will
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Syntax/GreenNode.cs:911]:
```suggestion
         * There is another overload for IReadOnlyList, since most collections already implement this, so checking for it will
```

**Review by sharwell — APPROVED** [review]

File: src/Compilers/Core/Portable/Syntax/GreenNode.cs [hunk 17]
```diff
@@ -932,16 +932,15 @@ private static bool EquivalentToInternal(GreenNode node1, GreenNode node2)
                     return SyntaxList.List(select(list[0]), select(list[1]));
                 case 3:
                     return SyntaxList.List(select(list[0]), select(list[1]), select(list[2]));
-                default:
-                {
+                default: {
```
Comment (333fred) [inline src/Compilers/Core/Portable/Syntax/GreenNode.cs:935]: Please put these on the next line and indent appropriately.

Comment (sharwell) [reply src/Compilers/Core/Portable/Syntax/GreenNode.cs:935]: 📝 The Correctness_Build failure is actually caused by this.

@HurricanKai if you ever want to run with full analyzers to see if there are warnings that could pop up during CI, you can use this command locally:

```
.\Build.cmd -a -rebuild
```

File: src/Compilers/Core/Portable/Syntax/GreenNode.cs (hunk 15 above)
Comment (HurricanKai) [reply src/Compilers/Core/Portable/Syntax/GreenNode.cs:944]: I also thought so. Note however that the only case where this is used is from WithLeading/Trailing trivia, and in those cases it seems more beneficial to check for IROList. See the List/Array benchmark above, it's not Span, but it indicates less special = better. And if I'm not missing anything ROSpan is a rare thing to be provided. (we could look into special casing for Array/Memory/ROMemory/Span/ROSpan and call into the ROSpan overload in all those cases. Will try and maybe open another PR)

Comment (sharwell) [reply src/Compilers/Core/Portable/Syntax/GreenNode.cs:944]: > And if I'm not missing anything ROSpan is a rare thing to be provided

`ReadOnlySpan<T>` cannot be cast to `IEnumerable<T>`, so to use this form all the callers would need to be updated to provide the correct type.

File: src/Compilers/Core/Portable/Syntax/SyntaxNodeOrTokenList.cs (hunk 12 above)
Comment (333fred) [reply src/Compilers/Core/Portable/Syntax/SyntaxNodeOrTokenList.cs]: @HurricanKai you'll want to either undo this one, or add a `RequiredUnderlyingNode` to `SyntaxNodeOrToken` like `SyntaxToken` has.

File: src/Compilers/Core/Portable/Syntax/GreenNode.cs (hunk 17 above)
Comment (HurricanKai) [reply src/Compilers/Core/Portable/Syntax/GreenNode.cs]: Thanks, `-rebuild` somehow tripped it up from time to time (sometimes the xlfUpdate task didn't really work, sometimes it complained about duplicate `-build`) but `-a -warnAsError` worked perfectly for me 😄

**Comment by HurricanKai** [comment]
Not sure why those tests failed. Seems like something timed out?

**Comment by 333fred** [comment]
Thanks @HurricanKai!
